### ISSUE TITLE
[BUG]/[FR]? Using a potomicrosensor reliably works for probing, but not in READ(Z_MIN_PROBE_PIN)

### ISSUE THREAD

**Issue by Sierra410** [MarlinFirmware/Marlin#25478]
### Did you test the latest `bugfix-2.1.x` code?

Yes, and the problem still exists.

### Bug Description

[I made an ABL probe](https://www.printables.com/model/417960-in-shroud-mounted-ender-3-bed-leveling-probe) which uses a photomicrosensor instead of a microswitch.  

While actual bed leveling works fine (the interrupts do fire), reading the state of the pin manually (`READ(Z_MIN_PROBE_PIN)`) doesn't produce reliable results, which causes the PROBE_TRIGGERED_WHEN_STOWED_TEST logic to fail most of the time.  

### Expected behavior

`READ(Z_MIN_PROBE_PIN)` returns proper state.

### Actual behavior

`READ(Z_MIN_PROBE_PIN)` returns correct state only sometimes.

### Version of Marlin Firmware

bugfix-2.1.x

### Printer model

Ender 3

### Electronics

Creality 4.2.2 motheboard

### Add-ons

https://www.printables.com/model/417960-in-shroud-mounted-ender-3-bed-leveling-probe

### Bed Leveling

ABL Bilinear mesh

### Your Slicer

None

### Host Software

None

### Don't forget to include

- [ ] A ZIP file containing your `Configuration.h` and `Configuration_adv.h`.

### Additional information & file uploads

I'm almost entirely sure that this is due to the fact that the sensor is an analog one and doesn't prevent the flow of current completely.  

While this could probably be fixed in hardware by adding a comparator, it seems that it could also be done in software.  

If I'm reading the documentation of STM32 right, the Z probe pin on Creality 4.2.2 is PB1, which appears to be capable of working in analog mode.  

Considering the fact that interrupts work fine already, it might make sense to add native support for probes that use analog sensors.  

For the time being I just made a quick and dirty workaround in probe.cpp.
```c++
  #if ENABLED(PROBE_TRIGGERED_WHEN_STOWED_NO_TEST)

    endstops.enable_z_probe(false);
    probe_specific_action(deploy);

  #elif ENABLED(PROBE_TRIGGERED_WHEN_STOWED_TEST)
```

**Comment by yet-another-average-joe** [comment]
The design you link cannot realiably work as is. You need a comparator as close to the opto as possible. The wiring is just catching EMI. It will not work from an analog input. Of course, you could code a comparator and a Schmitt trigger in the firmware, but this will not solve the EMI problem : it has to be removed from the source. Rule of thumb !

The Touch-MI works this way and uses a well known and ubiquitous "Arduino sensor" ; this one : https://www.amazon.com/Willwin-Measuring-Comparator-Optocoupler-Arduino/dp/B0776RHKB1/ref=sr_1_56?crid=2W0FZLMVXW4WB&keywords=optocoupler+module&qid=1678353934&sprefix=optocoupler+module%2Caps%2C173&sr=8-56
The Touch-MI is reliable, I tested it slightly more repeatable than a genuine BLTouch a few years ago. The Touch-MI has been supported since Marlin 1.8.x or 1.9.x . Don't trust the "australian guy that teaches tech on YT" : his video about this bed leveling sensor is a pile of crap.

![20230309_104108](https://user-images.githubusercontent.com/43514870/223984187-2406c893-9ff7-4b13-8d35-bf449f7c9a75.jpg)

Long story short : build your ABL sensor around this module, and tell Martlin to use the Touch-MI. You should get great results.

You could also wire the opto to a separate PCB embedding a comparator, using a shielded cable (as short as possible). The schematics for this module is all around the web.

**Comment by Sierra410** [comment]
Oh. I really didn't think of EMI. I feel pretty stupid now.  

I assumed since it uses a pullup resistor it shouldn't be affected by currents induced by EMI much.  But, I think what's going on is that the phototransistor in my sensor just fails to pull the pin down completely, letting EMI mess with things.  

I just slapped a 100K resistor between the probe pin and ground. High enough to keep it above logic level, low enough to help the sensor pull the pin to ground.  

I think standard deviation of 0.001575 with 0.006 of range for a probe that shouldn't work reliably is pretty good. I'm happy with that.  

Thanks for help, @yet-another-average-joe!

**Comment by yet-another-average-joe** [comment]
Just had a look to my notes about the BLTouch (just unboxed), dated may 2019.

BLTouch std dev : from 0.001 to 0.006 (ran 4 tests)

You're in the same ballpark. Just great results.

Numbers : 0.005506 , 0.001880 , 0.001027 , 0.001492

(this is repeatabilty of the sensor + repeatabilty of the machine ; the machine cannot be considered as perfect ; it is mind blowing)
